File: app/src/main/java/pl/pw/pkubik/smartalarm/MainActivity.java
package pl.pw.pkubik.smartalarm;

import android.app.Activity;
import android.app.AlarmManager;
import android.app.job.JobInfo;
import android.app.job.JobScheduler;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
import com.google.android.gms.common.GooglePlayServicesRepairableException;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.ui.PlacePicker;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;

import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getName();
    private static final int PLACE_PICKER_REQUEST = 1;
    private static final int TRAFFIC_JOB_ID = 1;
    private SettingsFragment settingsFragment;
    private AlarmManager alarmManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        settingsFragment = (SettingsFragment)
                getSupportFragmentManager().findFragmentById(R.id.activity_settings);

        alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
        scheduleTrafficJob();
    }

    public void launchPlacePicker(double latitude, double longitude) {
        Log.i(TAG, "Launching Place Picker intent.");
        PlacePicker.IntentBuilder builder = new PlacePicker.IntentBuilder();
        LatLngBounds bounds = LatLngBounds.builder()
                .include(new LatLng(latitude, longitude))
                .build();
        builder.setLatLngBounds(bounds);
        try {
            startActivityForResult(builder.build(this), PLACE_PICKER_REQUEST);
        } catch (GooglePlayServicesRepairableException |
                 GooglePlayServicesNotAvailableException e) {
            Toast.makeText(this, "Can't connect Google Play Services.", Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == PLACE_PICKER_REQUEST) {
            Log.i(TAG, "onActivityResult: PLACE_PICKER_REQUEST");
            if (resultCode == Activity.RESULT_OK) {
                Place place = PlacePicker.getPlace(this, data);
                settingsFragment.setPlace(place);
            } else {
                Log.i(TAG, "onActivityResult: User did not pick a place.");
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    public void scheduleTrafficJob() {
        Log.i(TAG, "scheduleTrafficJob: Scheduling a job for the next alarm.");
        AlarmManager.AlarmClockInfo nextAlarm = alarmManager.getNextAlarmClock();
        if (nextAlarm != null) {
            long time = nextAlarm.getTriggerTime() - TimeUnit.MINUTES.toMillis(5);
            JobScheduler scheduler = (JobScheduler) getSystemService(Context.JOB_SCHEDULER_SERVICE);

            // Cancel the previous job
            scheduler.cancel(TRAFFIC_JOB_ID);

            long timeDiff = time - System.currentTimeMillis();
            JobInfo jobInfo = new JobInfo.Builder(
                    TRAFFIC_JOB_ID, new ComponentName(this, TrafficJobService.class))
                    .setMinimumLatency(timeDiff)
                    .build();
            scheduler.schedule(jobInfo);
            settingsFragment.setAlarmTime(nextAlarm.getTriggerTime());
        } else {
            Toast.makeText(this, "No alarm clock exists", Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/pl/pw/pkubik/smartalarm/MainActivity.kt
package pl.pw.pkubik.smartalarm

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem

import com.google.android.gms.common.GooglePlayServicesNotAvailableException
import com.google.android.gms.common.GooglePlayServicesRepairableException
import com.google.android.gms.location.places.ui.PlacePicker
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.info
import org.jetbrains.anko.longToast


class MainActivity : Activity(), AnkoLogger {
    private lateinit var settingsFragment: SettingsFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        settingsFragment =
                fragmentManager.findFragmentById(R.id.activity_settings) as SettingsFragment
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onResume() {
        isRunning = true
        super.onResume()
    }

    override fun onPause() {
        isRunning = false
        super.onDestroy()
    }

    fun launchPlacePicker(id: Int, latitude: Double, longitude: Double) {
        info("Launching Place Picker intent.")
        val intentBuilder = PlacePicker.IntentBuilder()
        val bounds = LatLngBounds.builder()
                .include(LatLng(latitude, longitude))
                .build()
        intentBuilder.setLatLngBounds(bounds)
        try {
            startActivityForResult(intentBuilder.build(this), PLACE_PICKER_REQUEST + id)
        } catch (e: GooglePlayServicesRepairableException) {
            longToast("Can't connect Google Play Services.")
        } catch (e: GooglePlayServicesNotAvailableException) {
            longToast("Can't connect Google Play Services.")
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (requestCode >= PLACE_PICKER_REQUEST) {
            info("onActivityResult: PLACE_PICKER_REQUEST")
            if (resultCode == Activity.RESULT_OK) {
                val place = PlacePicker.getPlace(this, data)
                settingsFragment.setPlace(requestCode - PLACE_PICKER_REQUEST, place)
            } else {
                info("onActivityResult: User did not pick a place.")
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    fun scheduleTrafficJob(time: Long?) {
        info("scheduleTrafficJob: Scheduling a job for the next alarm.")
        val nextAlarm = Utils.getNextAlarm(this)
        if (nextAlarm != null) {
            longToast("Another alarm clock is already scheduled!")
        }

        if (time != null) {
            Utils.scheduleTrafficJob(this, time)
        } else {
            error("Received null as the time")
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        val id = item.itemId
        if (id == R.id.preview_option) {
            Utils.runAlarm(this)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        private val PLACE_PICKER_REQUEST = 10
        var isRunning = false
    }
}
